bcda/auth/oktabackend.go

```diff
@@ -0,0 +1,305 @@
+package auth
+
+import (
+	"bytes"
+	"crypto/rsa"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"math/big"
+	"net/http"
+	"net/url"
+	"os"
+	"sync"
+	"time"
+)
+
+var ob *OB
+var publicKeys map[string]rsa.PublicKey
+var once sync.Once
+
+var apiBase string
+var authString string
+var authServerID string
+
+type OB struct{}
+
+func NewOB() *OB {
+	once.Do(func() {
+		ob = &OB{}
+		publicKeys = getPublicKeys()
+		apiBase = os.Getenv("OKTA_CLIENT_ORGURL")
+		authServerID = os.Getenv("OKTA_OAUTH_SERVER_ID")
+		authString = fmt.Sprintf("SSWS %s", os.Getenv("OKTA_CLIENT_TOKEN"))
+		go refreshKeys()
+	})
+	return ob
+}
+
+func (ob *OB) PublicKeyFor(id string) (rsa.PublicKey, bool) {
+	key, ok := publicKeys[id]
+	log.Printf("invalid key id %s presented", id)
+	return key, ok
+}
+
+func (ob *OB) AddClientApplication(localID string) (string, string, error) {
+	body := fmt.Sprintf(`{ "client_name": "%s", "client_uri": null, "logo_uri": null, "application_type": "service", "redirect_uris": [], "response_types": [ "token" ], "grant_types": [ "client_credentials" ], "token_endpoint_auth_method": "client_secret_basic" }`, localID)
+	req, err := http.NewRequest("POST", apiBase+"/oauth2/v1/clients", bytes.NewBufferString(body))
+	if err != nil {
+		return "", "", err
+	}
+
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("Accept", "application/json")
+	req.Header.Add("Authorization", authString)
+
+	var client = &http.Client{Timeout: time.Second * 10,}
+	resp, err := client.Do(req)
+	if err != nil {
+		return "", "", err
+	}
+
+	if resp.StatusCode != 201 {
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			log.Printf("failed to create client %s because %v", localID, err)
```

Comment:
I think it's only error messages that shouldn't be complete sentences (capitalized + with punctuation). Not necessary for log messages.